x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -331,12 +359,133 @@ public String toString() {
+                                    final CachedBlobContainerIndexInput cachedIndexInput = (CachedBlobContainerIndexInput) input.clone();
+                                    assert loaded == false : "snapshot should not be fully loaded until all prewarming tasks are completed";
+
+                                    final int bytesRead = cachedIndexInput.prefetchPart(part); // TODO does not include any rate limitation
+                                    assert bytesRead == file.partBytes(part);
+
+                                    logger.trace(
+                                        () -> new ParameterizedMessage(
+                                            "{} part [{}/{}] of length [{}] prewarmed in cache for file [{}] in [{}] ms",
+                                            shardId,
+                                            part,
+                                            numberOfParts,
+                                            file.partBytes(part),
+                                            fileName,
+                                            TimeUnit.NANOSECONDS.toMillis(statsCurrentTimeNanosSupplier.getAsLong() - partStartTimeInNanos)
+                                        )
+                                    );
+                                } finally {
+                                    if (countDown.countDown()) {
+                                        IOUtils.closeWhileHandlingException(input);
+                                        logger.debug(
+                                            "{} cache prewarmed for file [{}] in [{}] ms",
+                                            shardId,
+                                            fileName,
+                                            TimeUnit.NANOSECONDS.toMillis(statsCurrentTimeNanosSupplier.getAsLong() - startTimeInNanos)
+                                        );
+                                    }
+                                }
+                            });
+                        }
+                    } catch (Exception e) {
+                        logger.warn(() -> new ParameterizedMessage("{} failed to prewarm cache for file [{}]", shardId, fileName), e);
+                    }
+                }
+            }
+        }
+
+        final String threadPoolName = ThreadPool.Names.SNAPSHOT;
+        final Executor executor = threadPool.executor(threadPoolName);
+        final int workers = Math.min(threadPool.info(threadPoolName).getMax(), queue.size());
+
+        final CountDownLatch latch = new CountDownLatch(workers);
+        for (int i = 0; i < workers; ++i) {
+            executor.execute(new AbstractRunnable() {
+                @Override
+                protected void doRun() throws Exception {
+                    CheckedRunnable<Exception> loader;
+                    while (isOpen && (loader = queue.poll(0L, TimeUnit.MILLISECONDS)) != null) {
```

Comment:
I see that this is taking the same approach as we use for uploading snapshot files (BlobStoreRepository). I would prefer not to hold onto workers for such a long time, as it can block the snapshot thread pool for a long time (cc: @original-brownbear).
In both cases (also the one BlobStoreRepository), I would prefer for the worker to process one file, then enqueue another task to the thread pool to pick up the next piece of work. This allows other operations to make progress as well, instead of waiting for a long time in the snapshot queue.

Reply:
I agree, that makes sense Yannick.

Reply:
I think for the uploading side the downside of

>  This allows other operations to make progress as well

is that it causes the index commits to be held on for a suboptimally long time. That's why the approach of fully monopolizing the pool was consciously chosen for uploads there.

Reply:
That's something to be better controlled at the `SnapshotShardsService` level then, though. It could limit the number of shards to be snapshotted concurrently by lazily enqueuing there.